File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/SearchRowPolicy.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.log.LogLevel;
import com.yahoo.messagebus.ErrorCode;
import com.yahoo.messagebus.Reply;
import com.yahoo.messagebus.metrics.MetricSet;
import com.yahoo.messagebus.routing.RoutingContext;
import com.yahoo.messagebus.routing.RoutingNodeIterator;

import java.util.HashSet;
import java.util.Set;
import java.util.logging.Logger;

/**
 * @author <a href="mailto:simon@yahoo-inc.com">Simon Thoresen</a>
 */
public class SearchRowPolicy implements DocumentProtocolRoutingPolicy {

    private static Logger log = Logger.getLogger(SearchRowPolicy.class.getName());
    private int minOk = 0; // Hide OUT_OF_SERVICE as long as this number of replies are something else.

    /**
     * Creates a search row policy that wraps the underlying search group policy in case the parameter is something
     * other than an empty string.
     *
     * @param param The number of minimum non-OOS replies that this policy requires.
     */
    public SearchRowPolicy(String param) {
        if (param != null && param.length() > 0) {
            try {
                minOk = Integer.parseInt(param);
            }
            catch (NumberFormatException e) {
                log.log(LogLevel.WARNING, "Parameter '" + param + "' could not be parsed as an integer.", e);
            }
            if (minOk <= 0) {
                log.log(LogLevel.WARNING, "Ignoring a request to set the minimum number of OK replies to " + minOk + " " +
                                          "because it makes no sense. This routing policy will not allow any recipient " +
                                          "to be out of service.");
            }
        }
    }

    @Override
    public void select(RoutingContext context) {
        context.addChildren(context.getMatchedRecipients());
        context.setSelectOnRetry(false);
        if (minOk > 0) {
            context.addConsumableError(ErrorCode.SERVICE_OOS);
        }
    }

    @Override
    public void merge(RoutingContext context) {
        if (minOk > 0) {
            Set<Integer> oosReplies = new HashSet<Integer>();
            int idx = 0;
            for (RoutingNodeIterator it = context.getChildIterator();
                 it.isValid(); it.next())
            {
                Reply ref = it.getReplyRef();
                if (ref.hasErrors() && DocumentProtocol.hasOnlyErrorsOfType(ref, ErrorCode.SERVICE_OOS)) {
                    oosReplies.add(idx);
                }
                ++idx;
            }
            if (context.getNumChildren() - oosReplies.size() >= minOk) {
                DocumentProtocol.merge(context, oosReplies);
                return;
            }
        }
        DocumentProtocol.merge(context);
    }

    @Override
    public void destroy() {
        // empty
    }

    @Override
    public MetricSet getMetrics() {
        return null;
    }
}

File: documentapi/src/vespa/documentapi/messagebus/policies/searchrowpolicy.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "searchrowpolicy.h"
#include <vespa/documentapi/messagebus/documentprotocol.h>

#include <vespa/log/log.h>
LOG_SETUP(".searchrowpolicy");

namespace documentapi {

SearchRowPolicy::SearchRowPolicy(const string &param) :
    _minOk(0)
{
    if (param.length() > 0) {
        int minOk = atoi(param.c_str());
        if (minOk > 0) {
            _minOk = (uint32_t)minOk;
        } else {
            LOG(warning,
                "Ignoring a request to set the minimum number of OK replies to %d because it makes no sense. "
                "This routing policy will not allow any recipient to be out of service.", minOk);
        }
    }
}

SearchRowPolicy::~SearchRowPolicy() {}

void
SearchRowPolicy::select(mbus::RoutingContext &context)
{
    std::vector<mbus::Route> recipients;
    context.getMatchedRecipients(recipients);
    context.addChildren(recipients);
    context.setSelectOnRetry(false);
    if (_minOk > 0) {
        context.addConsumableError(mbus::ErrorCode::SERVICE_OOS);
    }
}

void
SearchRowPolicy::merge(mbus::RoutingContext &context)
{
    if (_minOk > 0) {
        std::set<uint32_t> oosReplies;
        uint32_t idx = 0;
        for (mbus::RoutingNodeIterator it = context.getChildIterator();
             it.isValid(); it.next())
        {
            const mbus::Reply &ref = it.getReplyRef();
            if (ref.hasErrors() && DocumentProtocol::hasOnlyErrorsOfType(ref, mbus::ErrorCode::SERVICE_OOS)) {
                oosReplies.insert(idx);
            }
            ++idx;
        }
        if (context.getNumChildren() - oosReplies.size() >= _minOk) {
            DocumentProtocol::merge(context, oosReplies);
            return;
        }
    }
    DocumentProtocol::merge(context);
}

}
